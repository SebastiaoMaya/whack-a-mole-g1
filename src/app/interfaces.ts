type UUID = string;

export interface Player {
  id: UUID;
  name: string;
  score: number;
}
