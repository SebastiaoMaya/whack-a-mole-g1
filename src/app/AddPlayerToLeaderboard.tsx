import { Button, Flex, Input, Text } from "@chakra-ui/react";
import React, { useCallback, useMemo, useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import { getPlayerName, getScore } from "store/selectors";
import { addPlayerToLeaderboard, changePlayerName } from "store/gameSlice";
import { v4 as uuidv4 } from "uuid";
import { Player } from "./interfaces";
import { addPlayerToLeaderboardApi } from "network/api";

export const AddPlayerToLeaderboard = () => {
  const initPlayerName = useSelector(getPlayerName);
  const [playerName, setPlayerName] = useState(initPlayerName);
  const [isSubmitted, setIsSubmitted] = useState(false);

  const currentScore = useSelector(getScore);
  const dispatch = useDispatch();

  const handlePlayerNameChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const name = e.target.value;
    setPlayerName(name);
    dispatch(changePlayerName(name));
  };

  const player: Player = useMemo(
    () => ({ id: uuidv4(), name: playerName, score: currentScore }),
    [currentScore, playerName]
  );

  const handleOnClick = useCallback(async () => {
    try {
      await addPlayerToLeaderboardApi(player);
      dispatch(addPlayerToLeaderboard(player));
      setIsSubmitted(true);
    } catch (e) {
      console.error(e);
    }
  }, [dispatch, player]);

  if (isSubmitted) {
    return <Text>You are now on the Leaderboard!</Text>;
  }

  return (
    <Flex direction="column" mt="4">
      <Text>Congratulations, you are on the leaderboard!</Text>
      <Flex direction="row" alignItems="center" mt="4">
        <Input
          color="black"
          value={playerName}
          onChange={handlePlayerNameChange}
          placeholder="Type your name"
          height="100%"
          width="100%"
        />
        <Button background="green" px="3" py="2" onClick={handleOnClick}>
          Submit
        </Button>
      </Flex>
    </Flex>
  );
};
