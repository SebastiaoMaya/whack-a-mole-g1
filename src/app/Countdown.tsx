import { useDispatch, useSelector } from "react-redux";
import { useInterval } from "usehooks-ts";
import { decrementTimer, endGame } from "store/gameSlice";
import { getIsPlaying, getTimer } from "store/selectors";
import { Text } from "@chakra-ui/react";
import { useMemo } from "react";
import styled from "@emotion/styled";

interface StylingProps {
  isAlmostFinished: boolean;
}

const StyledText = styled(Text, {
  shouldForwardProp: (prop) => prop !== "isAlmostFinished",
})<StylingProps>`
  color: ${({ isAlmostFinished }) => (isAlmostFinished ? "red" : "black")};
`;

export const Countdown = () => {
  const dispatch = useDispatch();
  const isPlaying = useSelector(getIsPlaying);
  const timer = useSelector(getTimer);

  const minutes = useMemo(() => {
    const mins = Math.floor(timer / 60);
    if (mins < 10) {
      return "0" + mins;
    }
    return mins.toString();
  }, [timer]);

  const seconds = useMemo(() => {
    const secs = timer % 60;
    if (secs < 10) {
      return "0" + secs;
    }
    return secs.toString();
  }, [timer]);

  const tick = () => {
    if (timer > 0) {
      dispatch(decrementTimer());
    }
    if (timer === 0) {
      dispatch(endGame());
    }
  };

  useInterval(tick, isPlaying ? 1000 : null);

  return (
    <StyledText fontSize="2xl" fontWeight="bold" isAlmostFinished={timer < 10}>
      {minutes}:{seconds}
    </StyledText>
  );
};
