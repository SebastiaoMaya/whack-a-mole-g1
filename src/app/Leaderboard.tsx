import { Flex, Text } from "@chakra-ui/react";
import { useSelector } from "react-redux";
import { getLeaderboard, getScore } from "../store/selectors";
import { LeaderboardPlayer } from "./LeaderboardPlayer";
import { useMemo } from "react";

export const Leaderboard = () => {
  const leaderboard = useSelector(getLeaderboard);
  const currentScore = useSelector(getScore);
  const top10 = useMemo(() => leaderboard.slice(0, 10), [leaderboard]);

  //const currentPlayerLeaderBoardPosition = useMemo(()=>, [currentScore, top10])

  return (
    <Flex
      height="100vh"
      justifyContent="center"
      zIndex="1"
      opacity="0.7"
      background="black"
      position="absolute"
      top="0"
      width="100%"
      padding="3rem"
    >
      <Flex direction="column" color="white" alignItems="space-between">
        <Text fontSize="3xl" pb="4">
          Top 10 Players
        </Text>
        {top10 &&
          top10.map(({ id, ...rest }, idx) => (
            <LeaderboardPlayer
              key={`leaderboard_player_${id}`}
              position={idx + 1}
              {...rest}
            />
          ))}
      </Flex>
    </Flex>
  );
};
