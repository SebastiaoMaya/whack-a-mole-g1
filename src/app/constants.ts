export const COLUMNS = 4;
export const ROWS = 3;

export const MIN_VISIBILITY_DURATION_MS = 500;
export const MAX_VISIBILITY_DURATION_MS = 1000;

export const MIN_UNDERGROUND_DURATION_MS = 1000;
export const MAX_UNDERGROUND_DURATION_MS = 10000;
