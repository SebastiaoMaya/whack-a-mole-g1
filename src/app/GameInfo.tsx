import { Flex } from "@chakra-ui/react";
import { Score } from "./Score";
import { Countdown } from "./Countdown";
import { StartGameBtn } from "./StartGameBtn";

export const GameInfo = () => (
  <Flex justifyContent="space-between">
    <Score />
    <Countdown />
    <StartGameBtn />
  </Flex>
);
