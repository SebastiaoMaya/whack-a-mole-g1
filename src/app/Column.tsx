import { Flex } from "@chakra-ui/react";
import { Mole } from "./Mole";
import { ROWS } from "./constants";
import { v4 as uuidv4 } from "uuid";

export const Column = () => {
  return (
    <Flex direction="column" gap="10px">
      {[...Array(ROWS)].map(() => (
        <Mole key={`mole_${uuidv4()}`} />
      ))}
    </Flex>
  );
};
