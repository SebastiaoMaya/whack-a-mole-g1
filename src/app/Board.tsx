import { Flex } from "@chakra-ui/react";
import { Column } from "./Column";
import { COLUMNS } from "./constants";
import { Score } from "./Score";
import { v4 as uuidv4 } from "uuid";

export const Board = () => {
  return (
    <Flex
      backgroundImage="url('/WAM_bg.jpg')"
      backgroundPosition="center"
      backgroundRepeat="no-repeat"
      backgroundSize="cover"
      height="100vh"
      padding="20px"
      sx={{
        cursor: "url('/WAM_Hammer_scaled.png'), auto",
      }}
    >
      <Score />
      <Flex justifyContent="center" alignItems="center" gap="20px" width="100%">
        {[...Array(COLUMNS)].map(() => (
          <Column key={`column_${uuidv4()}`} />
        ))}
      </Flex>
    </Flex>
  );
};
