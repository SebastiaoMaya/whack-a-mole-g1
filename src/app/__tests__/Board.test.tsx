import { AppStore, setupStore } from "store";
import { GameSliceState } from "store/gameSlice";
import { renderWithProviders } from "utils/test-utils";
import { fireEvent, screen } from "@testing-library/react";
import { Board } from "app/Board";

describe("Board", () => {
  let store: AppStore;

  beforeAll(() => {
    store = setupStore();
  });

  it("matches Board snapshot", async () => {
    const { container } = renderWithProviders(<Board />, { store });
    expect(container).toMatchSnapshot();
  });

  it("clicks retry button and removes overlay", async () => {
    const initialGameSlice: GameSliceState = {
      playerName: "anonymous",
      score: 0,
      isPlaying: false,
      timer: 120,
      leaderboard: [],
      gameFinished: true,
    };

    renderWithProviders(<Board />, {
      preloadedState: {
        game: initialGameSlice,
      },
    });

    expect(screen.getByText("Top 10 Players")).toBeInTheDocument();

    const retry = screen.getByRole("button");
    fireEvent.click(retry);

    expect(screen.queryByText("Top 10 Players")).toBeNull();
  });
});
