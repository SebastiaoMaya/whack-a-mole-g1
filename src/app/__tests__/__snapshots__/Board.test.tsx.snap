// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Board matches Board snapshot 1`] = `
<div>
  <div
    class="css-9y6172"
  >
    <div
      class="css-1ynanbq"
    >
      <div
        class="css-gg4vpm"
      >
        <div
          class="css-dvxtzn"
        >
          <p
            class="chakra-text css-y0zjdx"
          >
            Score
          </p>
          <p
            class="chakra-text css-mbmqby"
          >
            0
          </p>
        </div>
        <p
          class="chakra-text css-1ri06s6"
        >
          02
          :
          00
        </p>
        <button
          class="chakra-button css-s7apec"
          role="button"
          type="button"
        >
          <p
            class="chakra-text css-1hgmzub"
          >
            Start
          </p>
        </button>
      </div>
      <div
        class="css-1fltl10"
      >
        <div
          class="css-eecy6w"
          draggable="false"
        >
          <div
            class="css-1mm82d"
          >
            <img
              class="css-68z3eg"
              draggable="false"
              src="/WAM_Hole.png"
            />
          </div>
          <div
            class="css-1mm82d"
          >
            <img
              class="css-68z3eg"
              draggable="false"
              src="/WAM_Hole.png"
            />
          </div>
          <div
            class="css-1mm82d"
          >
            <img
              class="css-68z3eg"
              draggable="false"
              src="/WAM_Hole.png"
            />
          </div>
        </div>
        <div
          class="css-eecy6w"
          draggable="false"
        >
          <div
            class="css-1mm82d"
          >
            <img
              class="css-68z3eg"
              draggable="false"
              src="/WAM_Hole.png"
            />
          </div>
          <div
            class="css-1mm82d"
          >
            <img
              class="css-68z3eg"
              draggable="false"
              src="/WAM_Hole.png"
            />
          </div>
          <div
            class="css-1mm82d"
          >
            <img
              class="css-68z3eg"
              draggable="false"
              src="/WAM_Hole.png"
            />
          </div>
        </div>
        <div
          class="css-eecy6w"
          draggable="false"
        >
          <div
            class="css-1mm82d"
          >
            <img
              class="css-68z3eg"
              draggable="false"
              src="/WAM_Hole.png"
            />
          </div>
          <div
            class="css-1mm82d"
          >
            <img
              class="css-68z3eg"
              draggable="false"
              src="/WAM_Hole.png"
            />
          </div>
          <div
            class="css-1mm82d"
          >
            <img
              class="css-68z3eg"
              draggable="false"
              src="/WAM_Hole.png"
            />
          </div>
        </div>
        <div
          class="css-eecy6w"
          draggable="false"
        >
          <div
            class="css-1mm82d"
          >
            <img
              class="css-68z3eg"
              draggable="false"
              src="/WAM_Hole.png"
            />
          </div>
          <div
            class="css-1mm82d"
          >
            <img
              class="css-68z3eg"
              draggable="false"
              src="/WAM_Hole.png"
            />
          </div>
          <div
            class="css-1mm82d"
          >
            <img
              class="css-68z3eg"
              draggable="false"
              src="/WAM_Hole.png"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
